### DISCUSSION TITLE
[RFC] Proposal: Support views, materialized views and foreign tables

### DISCUSSION THREAD

**Discussion by arnaud** [directus/directus#17265] [Feature Requests]
> In this proposal, any mention of "views" will reference "views, materialized views and foreign tables", as long as the
> vendors support it.

## Goal

The general idea is to support views, materialized views and foreign tables as first-class citizens in Directus, and be able to create, query, and CRUD them from Directus.

## Features

- **Discoverability for pre-existing views:** Directus should let me configure an already-existing SQL view;
- **Create views with SQL:** Create views from the admin app with custom-crafted SQL and a live preview;
- **Create views in no-code:** Create views from the admin app with a query builder and a live preview (with [metabase](https://www.metabase.com/docs/latest/questions/query-builder/introduction#the-query-builder) as an inspiration);
- **Edit view collections:** Views are editable as long as the data is mapped with unique ids;
- **Calculated fields:** Implicit support for calculated fields (fields with values based on other fields);

## Use cases

- Manage permissions;
- Merge data from multiple related tables;
- Use views with calculated fields;
- Thanks to foreign table support:
  - Use Directus as a data aggregation platform, with some help of [foreign-data wrappers](https://wiki.postgresql.org/wiki/Foreign_data_wrappers);
  - Enable some git-based workflows with Directus;

## Requirements

- Vendor agnostic
- Documented
- No-code
- Integrates well with existing features (admin app, API, flows, dashboards)

## Approach

- Currently, only tables are collections; Views would be collections too;
- Implement Views support into Knex itself, as per [@rijkvanzanten](https://github.com/directus/directus/discussions/3097#discussioncomment-322416);
- Declare/configure vendor support at knex-level in the [schema package](https://github.com/directus/directus/tree/main/packages/schema/src/dialects);
- Primary key identification:
  - Handle the presence of an "id" field as the PK;
  - On-the-fly PK generation for views with `SELECT ROW_NUMBER() OVER() AS id`;
  - Make the user choose the PK from available fields;

### Definitions

- _View_ : view of a query. The view is not physically materialized. Instead, the query is run every time the view is referenced in a query. ([reference](https://www.postgresql.org/docs/15/sql-createview.html#id-1.9.3.97.5))
- _Materialized View_ : materialized view of a query. The query is executed and used to populate the view at the time the command is issued and may be refreshed later. ([reference](https://www.postgresql.org/docs/15/sql-creatematerializedview.html#id-1.9.3.71.5))
- _Foreign Table_ : A foreign table references an external data, either from the table of another remote database, or a different schema, from a remote API or even a parseable local file. ([reference](https://www.postgresql.org/docs/15/sql-createforeigntable.html#SQL-CREATEFOREIGNTABLE-DESCRIPTION))

### Vendors support

| Vendor        | Views | Materialized views | Foreign tables |
| ------------- | :---: | :----------------: | :------------: |
| `postgres`    |   ✔   |         ✔          |       ✔        |
| `mysql`       |   ✔   |                    |                |
| `cockroachdb` |   ✔   |         ✔          |                |
| `sqlite3`     |   ✔   |                    |                |
| `oracledb`    |   ✔   |         ✔          |                |
| `mssql`       |   ✔   |         ✔          |                |

## Proof of concept

| Views discoverability | Create views with SQL | Live preview the view definition |
|-|-|-|
| ![directus-20230116-support-for-views](https://user-images.githubusercontent.com/44078/212573652-3013e173-d263-4f93-9fd8-066e50c66fce.gif) | ![directus-20230117-create-views](https://user-images.githubusercontent.com/44078/212786443-c0ca2592-ee9e-4013-bb03-216eaed7236b.gif) | ![directus-20230118-preview-views](https://user-images.githubusercontent.com/44078/213036041-08168813-e6ed-4c16-91de-f90e5b288db8.gif) |

### Try it out

Support for postgres views has been implemented in the following [custom Docker image](https://hub.docker.com/r/codename/directus/tags):

```shell
docker pull codename/directus:9.22.4-views-beta.8
```

> ⚠️ This is a proof of concept under heavy [ongoing development](https://github.com/arnaud/directus/tree/feat/views-support); Always be cautious with unofficial images.

---

- https://github.com/directus/directus/pull/17174
- #3097
- 💬 [**Discord**](https://discord.com/channels/725371605378924594/1064154620663955476)

**Comment by rijkvanzanten** [comment]
Dig this! The only thing I'm not on board with here is the reliance on raw SQL to create the views. Even though it gives the most flexibility, it also opens up a pretty nasty attack vector[^1], and means that it's very easy to add parts to a view that may or may not be supported by Directus (yet). It also means that the exact way to configure a view is different between various vendors, which makes documentation very tricky. It also means that we expect the user to be intimately familiar with SQL, which goes against the no-code/low-code approach of the app. I'm wondering how we can rely on a more user friendly UX that configures this through the API methods rather than SQL. (With the knowledge that Directus will still rely on mirroring of the DB, so advanced usage can be achieved by running the SQL directly in-DB)

[^1]: Even though data model is admin-only, an Admin in Directus isn't necessarily a sys-admin with root access to the database. Even if you are somehow able to elevate your permissions to the admin role through another vulnerability, you're still bound to what the Directus API can do to the database. When we add a submit-raw-sql-which-is-executed endpoint like above, it opens up the opportunity for admins to do massive damage, as they now have way more permissions to the DB. There's also no way for us to make sure an admin user doesn't execute anything but the creation of a View. Somebody could enter a `DROP DATABASE x;` and it'll happily execute thinking a view will be previewed.

**Reply by bamboowonder** [reply]
Maybe I'd add a perspective.  That for someone like myself, an admin in a small company.  I do have root on the db, and have set the directus user that accesses the db to be very limited to only what it should be allowed to do.  so "drop db" etc wouldn't work.   
May I proposed that sql access be possible perhaps as a very special permission or env variable?  In the similar vein as the api access code in the UI is well protected?
I imagine there may be many users and use-cases where the admin who configures the UI in directus, also has root on the db.  And of course there are many where allowing sql access to even the directus admin would be a no-go.

Or perhaps it could be set up as an extension, with big warnings?    Or maybe the raw sql is sandboxed/neutered/moderated such that sql like drop db or other nastiness are sent to /dev/null and simply aren't possible?   I agree that giving a naive admin a footgun isn't a good plan.  But dear goodness it would be such a time saver to write sql vs messing around with api calls.

And with all that said, I FULLY agree with the original proposal, that doing this in stages is both pragmatic, more secure and sane.   Simply starting with being able to SEE and interact with read-only views in the DB would be a massive improvement, and present limited issues.   Can we start with that?
And then have a phase 2 or 3 that slowly adds some of the more complex(many considerations) features later? 
Raw SQL access is a nice-to-have that could be a phase 4 type of thing.
Cheers

**Reply by rijkvanzanten** [reply]
> That for someone like myself, an admin in a small company. I do have root on the db [...] I agree that giving a naive admin a footgun isn't a good plan. But dear goodness it would be such a time saver to write sql vs messing around with api calls. [...]

Yeah! That's what I meant when I talked about the fact that Directus should still mirror the database. For users like yourself, I want to make sure that you can just login to the database, and do it however you want without having to faff about in the API. 🙂 

> And with all that said, I FULLY agree with the original proposal, that doing this in stages is both pragmatic, more secure and sane. Simply starting with being able to SEE and interact with read-only views in the DB would be a massive improvement, and present limited issues. Can we start with that?

From an implementation / PR perspective, yes! As a product release, I'm not super sure yet. Releasing it as a read-only implementation locks it away behind the aforementioned database knowledge and access requirement, which locks out a lot of users. Not ideal!

**Reply by claytongulick** [reply]
> Releasing it as a read-only implementation locks it away behind the aforementioned database knowledge and access requirement, which locks out a lot of users. Not ideal!

Not ideal, but better than no users having access to views at all.

We still need and use views. Directus' query DSL is great, but limited to fairly basic queries. Not implementing read-only views just means that it takes *even more* technical skill to achieve, because we make the view and a custom endpoint to access it, and have to roll custom permissions code, parameters, etc... In the endpoint.

**Reply by ctholho** [reply]
> It also means that the exact way to configure a view is different between various vendors, which makes documentation very tricky. It also means that we expect the user to be intimately familiar with SQL, which goes against the no-code/low-code approach of the app.

While I understand that this is a great overall strategy, one arguments against: Don't let perfect be the enemy of good. Couldn't you consider a continuous delivery approach here?

Regarding permissions: Shouldn't `knex.raw()` prevent totally arbitrary sql injection if you use it like in this stack overflow question: https://stackoverflow.com/questions/71581712/can-you-create-views-using-knex-js-defined-by-raw-sql

Regarding security overall: Directus already offers endpoints that deletes tables or even only single rows that could be crucial and wouldn't be so obvious as `drop db`. So misbehaving directus admins are already a big concern. `drop db` or more sneaky sql injections don't add much of a new attack vector and we should have anticipated it already.

> it's very easy to add parts to a view that may or may not be supported by Directus (yet)

If Directus mirrors the DB, than using direct access to the DB or creating a view with directus and raw SQL doesn't create any more footguns. Warning unsuspecting admins that this is an expert feature that could lead to having your database crash should be sufficient. To prevent denial of service or other likewise attacks you will likely need to set up tighter restrictions about long running, intensive queries. All of this gives you a splendid reason to offer flexible enterprise billing. Additionally, you could prohibit materialized views on the hosted directus if you worry about memory.

The great thing: Arbitrary SQL views would allow many more orgs to use your hosted directus. Restricting it to the API CRUD stuff would mean we have to use custom endpoints again.

**Reply by rijkvanzanten** [reply]
Thanks for your insight @claytongulick & @ctholho!

> While I understand that this is a great overall strategy, one arguments against: Don't let perfect be the enemy of good. Couldn't you consider a continuous delivery approach here?

Not quite sure what you mean by a "continuous delivery approach" in this context. Mind elaborating?

> Regarding permissions: Shouldn't knex.raw() prevent totally arbitrary sql injection if you use it like in this stack overflow question: https://stackoverflow.com/questions/71581712/can-you-create-views-using-knex-js-defined-by-raw-sql

It'll prevent injection attacks, yes, but it doesn't include a way to ensure that the command you're executing is a view creation command. For example, a user could run `CREATE VIEW` (expected), or `CREATE OR REPLACE VIEW` (still OK, but might erroneously overwrite something), or run `CREATE TABLE` (not good), or run `INSERT INTO test (stringColumn) VALUES (repeat("uh oh", 150000000))`[^1]. The parameterized query would mean that you can't SQL inject within the value or column name, but it doesn't prevent running the wrong query. For DBAs, this is not a problem cause it's obvious to not run the wrong query, but for the non-technical user pool, it becomes very easy to copy/paste the wrong query from the internet and get in serious trouble, which Directus then has no way of preventing or reverting, as it effectively bypasses the API. 

> Regarding security overall: Directus already offers endpoints that deletes tables or even only single rows that could be crucial and wouldn't be so obvious as drop db. So misbehaving directus admins are already a big concern. drop db or more sneaky sql injections don't add much of a new attack vector and we should have anticipated it already.

I disagree. Dropping tables, rows, or databases is only one of the many things you can do in SQL directly. There's tons of edge cases that will be exposed by running arbitrary SQL, like changing MySQL collations, installing PG extensions, changing Sqlite parameters for the connection, etc.

To me, this is where the direct-to-db angle comes in. For DBAs who know what they're doing and who are accepting those risks, they should be able to go straight to the database and run those queries anyways. In that case, it's also very obvious that you're bypassing Directus' security features.

> Warning unsuspecting admins that this is an expert feature that could lead to having your database crash should be sufficient.

I don't fully agree with this. There's no guarantee that an "Admin" user in Directus has (or should have) the same level of access to the underlying SQL database as the sysadmin does.

> The great thing: Arbitrary SQL views would allow many more orgs to use your hosted directus. Restricting it to the API CRUD stuff would mean we have to use custom endpoints again.

This is a good point that we're also solving in other ways. We're actively working on allowing Directus Cloud databases to be accessed directly (with separate firewall settings, like you'd expect to find on managed database services), and allowing those projects to connect to databases hosted elsewhere. So in the bigger discussion here around running SQL from Directus vs running SQL in the database, cloud vs self-hosted should be considered equal (although I do really appreciate you taking the monetized side into account to; gotta make money to run this somehow 😄) 👍🏻 

[^1]: Bit of an insane exaggerated example of course, but it does highlight the fact that a malicious bit of SQL that was copy pasted by an unexpecting admin user can do very malicious things.

**Reply by ctholho** [reply]
Hey, thanks for taking the time! 🎉

> Not quite sure what you mean by a "continuous delivery approach" in this context. Mind elaborating?

I'm talking about allowing read only views or other paths that could speed up views in directus. I'm referring to:

> From an implementation / PR perspective, yes! As a product release, I'm not super sure yet. Releasing it as a read-only implementation locks it away behind the aforementioned database knowledge and access requirement, which locks out a lot of users. Not ideal!

On to the next:
> I don't fully agree with this. There's no guarantee that an "Admin" user in Directus has (or should have) the same level of access to the underlying SQL database as the sysadmin does.

They don't have the same level of access. I wouldn't give a sysadmin role to the directus instance. The DB Admin can still limit the privileges they give to directus.

> It'll prevent injection attacks, yes, but it doesn't include a way to ensure that the command you're executing is a view creation command.

Oh, but `createViewOrReplace()` could be hard coded. You would only allow changing the `DAILY_SESSIONS_QUERY` ([as per the example](https://stackoverflow.com/questions/71581712/can-you-create-views-using-knex-js-defined-by-raw-sql))

> repeat("uh oh", 150000000))

This is what I mean with long running, expensive queries. If that is a concern, billing for resource usage could go hand in hand with that. But I feel that you don't need to go so overboard to protect admins from self hosted instances. They are already very dangerous for careless users. 

E.g. these things are easy to do with a lot of potential harm
* a mistaken permission insta-autosaved in the permissions panel
* accidentally deleting a table and clicking OK
* visibility / potential export of PII

Copy-pasting a malicious SQL view query would be the least of my concern as a directus admin myself.

Another tool that follows a similar approach regarding a low code GUI, [Metabase](https://www.metabase.com) – they allow admins to enable raw sql queries for all users or only admins.

**Reply by arnaud** [reply]
Just released a first working version with the following [custom Docker image](https://hub.docker.com/r/codename/directus/tags):

```shell
docker pull codename/directus:9.22.4-views-beta.8
```

This is a proof of concept, limited to only postgres views at the moment, with a basic yet working way of defining & previewing custom views directly from the Directus interface.

⚠️ Be warned that this implementation lacks basic security safeguards, as _database can be dropped_.

I hope that this release will push the most brave community members to come help with the tasks at hand, such as implementing a no-code interface, supporting all vendors and handling a tiny bit more of security. :)

**Reply by bamboowonder** [reply]
Although I'd love to help with coding, I don't have the chops for that.  But I can help with QA.      And this issue may not be caused by your fork, but it is illuminated by it.

We have many hundreds of Views and Materialized Views.  

In this fork, I notice that Directus has more of a lag in most action (page transitions, loading, saving etc).

Granted I have all caching / schema caching disabled as I'm doing active development.  I suspect that may be a culprit.

It seems odd though, that Directus would seem to want to be parsing the all tables/views for each page load.  

Wouldn't that be something only necessary on the Admin Data Model page? 

Anyhoo the solution is to exclude all the Views in the .env file down to only those I want Directus to manage.

And thank you for your efforts on this RFC!!

**Reply by claytongulick** [reply]
@rijkvanzanten totally get the concern with not wanting to allow DDL through the web UI. We don't want a Bobby Tables moment (https://xkcd.com/327/).

As a first step though, I'd love to see a maximally minimal approach where Directus was able to recognize existing views, keep them read-only, but treat them in other ways like a collection (permissions, api access, etc...).

It would be an enormous time saver and eliminate a whole class of security risks from rolling custom endpoints for each view we make. Just having declarative permissions on read-only views would be a massive step forward, imho.

**Reply by arnaud** [reply]
> Although I'd love to help with coding, I don't have the chops for that. But I can help with QA. And this issue may not be caused by your fork, but it is illuminated by it.
> 
> We have many hundreds of Views and Materialized Views.
> 
> In this fork, I notice that Directus has more of a lag in most action (page transitions, loading, saving etc).
> 
> Granted I have all caching / schema caching disabled as I'm doing active development. I suspect that may be a culprit.
> 
> It seems odd though, that Directus would seem to want to be parsing the all tables/views for each page load.
> 
> Wouldn't that be something only necessary on the Admin Data Model page?
> 
> Anyhoo the solution is to exclude all the Views in the .env file down to only those I want Directus to manage.
> 
> And thank you for your efforts on this RFC!!

As the directus implementation lets you choose from which schema to pull tables and views, I assume you should be able to split and rearrange the views into dedicated and separate schemas? Directus will only show those that belong to your Directus schema.

**Reply by bamboowonder** [reply]
Unfortunately it would break a tonne of legacy apps by moving all those 100s of views to a different schema just to accommodate Directus.  
Upon further investigation, by starting Directus with npx instead of systemd, after any schema change (add a field to a collection for instance), I watch the terminal fill up for nearly a minute as Directus parses *every* column of *every* view/table/mat.view, despite having listed them all in the excluded tables in the .env
Is that something that Directus does independently of your fork?  
I guess this is partly solved by letting Directus cache schema changes which I'll do after I'm done the heavy lifting of getting this implimentation off the ground.

**Reply by rijkvanzanten** [reply]
> @rijkvanzanten totally get the concern with not wanting to allow DDL through the web UI. We don't want a Bobby Tables moment (https://xkcd.com/327/).
> 
> As a first step though, I'd love to see a maximally minimal approach where Directus was able to recognize existing views, keep them read-only, but treat them in other ways like a collection (permissions, api access, etc...).
> 
> It would be an enormous time saver and eliminate a whole class of security risks from rolling custom endpoints for each view we make. Just having declarative permissions on read-only views would be a massive step forward, imho.

Totally! I think we might want to "split up" this feature request in two new ones. One for how Directus should behave when it comes to making views accessible read only, and one for how you can create/modify views from Directus

**Reply by claytongulick** [reply]
What sort of realistic appetite is there for this from the team, do you think?

I don't mind working on a PR, but I don't want to put all the time & effort in to have it languish, and we're not operating from a fork anymore so we wouldn't be able to use it if it wasn't merged.

**Comment by JonL1** [comment]
@rijkvanzanten I don't see any problem with creating a virtual collections based on raw SQL(or via JS for that matter). **At least for self-hosted.** 

You already allow extensions to be loaded and you also allow to run JS through the UI via the "Run script" operation which even if using vm/vm2 it is not bulletproof.

Both (extensions and "Run Script") are considered advanced features and both can be restricted in some way. This would be just another feature that needs good handling.

Also, it is something other "internal tools" competitors offer out-of-the-box.

**Reply by JonL1** [reply]
@rijkvanzanten also ENV variables are now accesible in the "Run Script" operation. I think we are already way passed the line of "someone with bad intentions or by mistake can do serious damage via directus".

In the end it's a matter of the directus admin role to prevent giving access to people that could break things and having a disaster recovery plan in place :)

Just as with any kind of production software.

**Reply by rijkvanzanten** [reply]
It's not that black and white. Modules and env var access in the run-script operation is controlled on an explicit allow-list basis configured in env, with nothing allowed by default.

**Reply by JonL1** [reply]
Same with collections, right? They are not allowed to be created by default. So a collection created from SQL would be restricted to admin role unless explicitly granted to other roles. Or am I missing something (probably I am)?

**Reply by AMB-Informacio** [reply]
We are very interested in being capable of showing views as collections. When talking about self-hosted Directus, I think it is less important to be able to create the views through it. In most cases you have acces to the DB and it's not a problem to create the view.

I think the problem is to fit this feature in Directus Cloud to make sense there, where you will need to allow views creation. I understand making this feature suitable for Directus Cloud is what will make this interesting to develop.

The key here, as @claytongulick already said: "As a first step though, I'd love to see a maximally minimal approach where Directus was able to recognize existing views, keep them read-only, but treat them in other ways like a collection (permissions, api access, etc...)."

To make it work, what I understood that has to be solved somehow is the need of an ID for each collection item. Is that right? If detail view is disabled can be ignored?

At this time, a workaround that can be valid in some cases is to create a table where to insert the results of querying the view. Like a kind of materialized view (or view cache). With a global table to maintain a state (dirty/stale/uptodate) for each of theses tables (views caches). Then with triggers make it dirty when changing a table/items the view depends upon (or maybe when an amount of time has passed), and "rehydrate" when wanting to view items (if it's dirty). You also can, adding a cache_ts field, even maintain versions of the view results!

I think that trying to make possible to create views, in a controlled way, with UI is too complex. But if the admin user have a spot to make queries directly to the underlying db system, with properly controlled permissions, then he could be able to create views.  A thing that is not performed in an operational basis.

**Comment by avimar** [comment]
>"As a first step though, I'd love to see a maximally minimal approach where Directus was able to recognize existing views, keep them read-only, but treat them in other ways like a collection (permissions, api access, etc...)."

Is there a blocker to manually creating a view in postgres and marking it all read-only in directus, and just using it for read-only?

(What is "mirroring" a collection I've seen referenced in various places?)

EDIT: I tried it - I made a collection inside directus then replaced it with the view inside postgres. But it shows up only as a folder or something inside directus. I assume the data-introspection realized it's not a real table and deleted the fields...

**Reply by BrianJM** [reply]
> manually creating a view in postgres and marking it all read-only in directus

I came here to see if this was a viable workaround.

**Comment by bamboowonder** [comment]
While I had great hopes for the View support, ultimately we dropped the fork we'd made that had this View support.   We needed production stability and without the hassle of merging updates.

Our workaround, is that we have triggers that drop and regenerate postgres "unlogged" tables.  And oddly enough, this has actually worked really well.  It is more or less the same as how a materialized view works, and has all the performance benefits versus tradition views (as the underlying query to build these is quite complex and resource intensive).   And for the most part, the bookmarks and saved queries in Directus are good enough that we've just phased out the extensive use of views we previously had.   

It's one of many examples of how we've re-architected our database design to function better with Directus, because Directus is such a good fit for our internal apps and db management.   It was worth it.

I'd recommend anyone stuck by the lack of view support to consider "unlogged tables" with a trigger or cronjob to regenerate them.

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

(Also ref https://github.com/directus/directus/discussions/3097 for more discussion on the same topic)

**Reply by ddelange** [reply]
Hi @rijkvanzanten 👋 

We're at 131 upvotes by now. Even read-only view support would already go a long way (and make this development effort more incremental). Can the team provide an update here?

**Comment by AMB-Informacio** [comment]
Thanks!!!

Daniel Valls Estella
Tècnic Especialista
Via Laietana, 33, 4t 1a - 08003 Barcelona
Tel. (+34) 93 291 80 96
***@***.***
www.ambinformacio.cat
La informació inclosa en aquest correu electrònic és CONFIDENCIAL, i és per a ús exclusiu del destinatari. Si vostè llegeix aquest missatge i no n'és el destinatari, li informem que està totalment prohibida la divulgació, distribució o reproducció d'aquesta comunicació, i li demanem que ens ho notifiqui i ens retorni aquest missatge, i a més que l'esborri. Gràcies.
Abans d'imprimir aquest e-mail pensi en el medi ambient.
La información incluida en este correo electrónico es CONFIDENCIAL, siendo para uso exclusivo del destinatario. Si Vd. lee este mensaje y no es el destinatario, le informamos de que está totalmente prohibida la divulgación, distribución o reproducción de este comunicado y le agradeceríamos lo notificara y lo devolviera a la dirección arriba indicada, borrando el mensaje original. Gracias.
Antes de imprimir este e-mail piense en el medio ambiente.
This email and the information within are CONFIDENTIAL and it is intended exclusively for the addressee. If this message has been received in error, you are not entitled to use, disclose, distribute copy or rely on this email in any way. Please notify us immediately by email and delete it from your system.
Please consider the environment before printing this email. 
________________________________
De: Rijk van Zanten ***@***.***>
Enviat el: dimecres, 10 de maig de 2023 22:41
Per a: directus/directus ***@***.***>
A/c: informatica ***@***.***>; Comment ***@***.***>
Tema: Re: [directus/directus] [RFC] Proposal: Support views, materialized views and foreign tables (Discussion #17265)


Heya! Thanks for opening this feature request!

This feature request has received over 15 votes from the community. This means we'll move this feature request to the Under Review state!

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

For more information, see our Feature Request Process<https://docs.directus.io/contributing/feature-request-process.html>.

(Also ref #3097<https://github.com/directus/directus/discussions/3097> for more discussion on the same topic)

**Comment by timokluser-dev** [comment]
Hi, is there an update on SQL Views for Directus? @rijkvanzanten

**Comment by lostation** [comment]
@arnaud What a great idea you had Thanks !!! @rijkvanzanten We absolutely want that powerful feature !!! I totally agree about breaking changes of a lot of legacy running apps...but if the extended new view feature could be only selected by .env selector or any ui toggler... Then we obsiously take the responsability of messing up the app or not and it won't break anything by default. For my case, I need to integrate an external foreign table from a distant server and I only want to list the content and join some rows with local directus entities. Is that sounds so terrible for you ? Thanks

**Reply by dorian100** [reply]
I have exactly the same needs, it would be great and very powerful for us. Thanks for this good idea !

**Comment by mabril** [comment]
We really need this to continue working with directus.

**Comment by chrisgoddard** [comment]
Hi! Any updates on this?

**Reply by arnaud** [reply]
Sadly, none on my end.
I suspect this matter needs more traction in order for the directus core team to address this challenge.

**Comment by AMB-Informacio** [comment]
I think we must split it up in, at least, two parts. The first one considering only the case in which the views are created outside Directus.

Then take care of collection view only, and not the detail view. Let collection items have no ID but restricting then the possibility of item creation or detail view.

I think it would be a good first step on that will let build upon it.

Enviat des de l'Outlook per a l'iOS<https://aka.ms/o0ukef>

Daniel Valls Estella
Tècnic Especialista
Via Laietana, 33, 4t 1a - 08003 Barcelona
Tel. (+34) 93 291 80 96
***@***.***
www.ambinformacio.cat
La informació inclosa en aquest correu electrònic és CONFIDENCIAL, i és per a ús exclusiu del destinatari. Si vostè llegeix aquest missatge i no n'és el destinatari, li informem que està totalment prohibida la divulgació, distribució o reproducció d'aquesta comunicació, i li demanem que ens ho notifiqui i ens retorni aquest missatge, i a més que l'esborri. Gràcies.
Abans d'imprimir aquest e-mail pensi en el medi ambient.
La información incluida en este correo electrónico es CONFIDENCIAL, siendo para uso exclusivo del destinatario. Si Vd. lee este mensaje y no es el destinatario, le informamos de que está totalmente prohibida la divulgación, distribución o reproducción de este comunicado y le agradeceríamos lo notificara y lo devolviera a la dirección arriba indicada, borrando el mensaje original. Gracias.
Antes de imprimir este e-mail piense en el medio ambiente.
This email and the information within are CONFIDENTIAL and it is intended exclusively for the addressee. If this message has been received in error, you are not entitled to use, disclose, distribute copy or rely on this email in any way. Please notify us immediately by email and delete it from your system.
Please consider the environment before printing this email. 
________________________________
De: Arnaud Leymet ***@***.***>
Enviat el: Wednesday, October 18, 2023 8:53:07 AM
Per a: directus/directus ***@***.***>
A/c: informatica ***@***.***>; Comment ***@***.***>
Tema: Re: [directus/directus] [RFC] Proposal: Support views, materialized views and foreign tables (Discussion #17265)


Sadly, none on my end.
I suspect this matter needs more traction in order for the directus core team to address this challenge.

**Comment by julian-hecker** [comment]
Hi all, I've been trying to get views in directus as it would enable some essential functionality for the use case I'm working on. I've created some views using raw SQL and tried everything I could to get it recognized as a collection, including modifying the directus_collections and directus_fields tables but wasn't able to get it to work. I get a 403 Forbidden error when trying to access the view's records. It would be great if Directus could at least recognize existing views. Has anyone had success with this?

**Reply by arnaud** [reply]
Yup https://github.com/directus/directus/discussions/17265#discussion-4780877

**Reply by bamboowonder** [reply]
While I really liked what @arnaud created, using it meant the added hassle of maintain our own fork, which just wasn't worth it.
The way we shoehorned around this, is writing some stored procedures in postgres that generate unlogged tables (basically the equivalent of a materialized view as far as performance etc).   and directus doesn't even notice when it is regenerated in the background.   some of the unlogged tables are generated using a cronjob, others are generated when an update to the underlying data occurs.   It was a pain to setup but it meant we could use the full functionality of directus without forks.     Not having proper view support has been a razors edge of us often reevaluating whether directus is worth it or not.

**Comment by w0kyj** [comment]
### Short Summary of Request Review April 18, 2024
Generally this is a very, very complex feature and discussion. We will work to break this into more manageable parts.
- Views are SQL specific and support varies by vendor
    - supporting views as a read-only concept makes sense
        - Things that must be considered:
            - data access permissions (view may aggregate data from multiple tables and the Directus user should/may not have access to all the data)
            - document security, permissions, and caveats for the Directus administrators
            - views would be created via direct SQL by the database system admin
            - restrict directus role permissions to Read configuration only Create/Update/Delete disabled
    - creating/updating views is not planned currently due to role/security concerns:
        - crosses boundary into system administration role
        - enables sql injection
- Discussed possibility of elevating `Bookmarks` to first class
    - bookmarks are a limited view feature already
    - adding ability to aggregate/join data/collections could be a way to implement Directus "views" as a normalized construct regardless of underlying `data source/vendor`

Lots of great discussion and as always feedback welcome!

**Reply by julian-hecker** [reply]
Didn't know about Bookmarks as a limited view feature. Could definitely see how it could be expanded upon to support view-like constructs.

**Comment by julian-hecker** [comment]
Any further discussion on this? Updates? It’s been quite a while.

**Comment by JamesW1** [comment]
Thanks for the suggestion! We've moved feature requests to our public roadmap at https://roadmap.directus.io/roadmap.

I've migrated this request there for you: https://roadmap.directus.io/roadmap?id=3b57654d-bd16-4b7b-b973-9a48c27f6c7d
